File: spring-cloud-integrated-demo/event-lib/src/main/java/com/github/jccode/springcloud/integrateddemo/common/event/Publisher.java
package com.github.jccode.springcloud.integrateddemo.common.event;


import org.springframework.stereotype.Component;

/**
 * Event Publisher
 *
 * 要点:
 *
 * 1. 需要先将事件持久化到DB, 待Kafka发送成功后, 在回调中更新发送状态为成功.
 *
 * 逻辑:
 * 1. 先将消息落地,写库;
 * 2. 定时器1 扫描状态为 New 的记录, 发布到 Kafka;
 * 3. 定时器2 扫描状态为 Pending 的记录, 重新发布到 Kafka.
 *
 */
@Component
public class Publisher {

    
}

File: spring-boot-kafka-demo/kafka-json-demo2/src/main/scala/com/github/jccode/springbootkafkademo/jsondemo22/event/Publisher.scala
package com.github.jccode.springbootkafkademo.jsondemo22.event

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.support.{KafkaHeaders, SendResult}
import org.springframework.messaging.Message
import org.springframework.messaging.support.MessageBuilder
import org.springframework.stereotype.Component
import org.springframework.util.concurrent.ListenableFuture

@Component
class Publisher {

  @Autowired
  private val template: KafkaTemplate[String, _] = null

  def send(message: Message[_]): ListenableFuture[SendResult[String, _]] = template.send(message)

  def send[T](topic: String, t: T): ListenableFuture[SendResult[String, _]] = {
    val message: Message[T] = MessageBuilder.withPayload(t).setHeader(KafkaHeaders.TOPIC, topic).build()
    send(message)
  }

  def send[T](topic: String, key: String, t: T): ListenableFuture[SendResult[String, _]] = {
    val message: Message[T] = MessageBuilder.withPayload(t).setHeader(KafkaHeaders.TOPIC, topic).setHeader(KafkaHeaders.MESSAGE_KEY, key).build()
    send(message)
  }

}
